chore: v1 api

pmengelbert opened the pull request:
Create v1 api

In order to create the v1 api for eraser, a number of additional changes had to be made.

Our eraser-tooling image had to be modified to include the `conversion-gen` tool. This is to generate type conversions between the older `v1alpha1` API and the newer `v1`.

To facilitate conversion between the two, an `unversioned` api was created. This becomes a common hub between `v1` and `v1alpha1` and facilitates conversion. It also makes it easier for our code: with the `unversioned` api, we no longer need to update large swaths of code when a new api version is introduced.

In order to get the `conversion-gen` tool to work, I needed to add several magic comments throughout the api code. Most importantly, I needed to add a `doc.go` file to each of the api version directories. `conversion-gen` looks for this file and will not work without it. Above the package name in those files, I added the following two magic comments:

```
+// +k8s:conversion-gen=github.com/Azure/eraser/api/unversioned
+// -external-types=github.com/Azure/eraser/api/v1
```

This tells `conversion-gen` to use the `unversioned` api as the conversion hub.

To skip kubebuilder processing the `unversioned` api, `+kubebuilder:skip` had to be added before the package declaration. To specify which api version should be stored in etcd, the `+kubebuilder:storageversion` comment was added to the `v1` types. In addition, `kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated"` was added to the `v1alpha1` API in order to warn the user to upgrade.

In the `groupversion_info.go` file of each version directory, I had to add the line

```go
localSchemeBuilder = runtime.NewSchemeBuilder(SchemeBuilder.AddToScheme)
```

To be perfectly honest, I'm not sure what this does, but it is required for the `conversion-gen` process to generate working code.

Finally, a new file was added to `test/e2e/test-data` which deploys an imagelist using the new API version. One of the tests was updated to use this file instead.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Create v1 api

In order to create the v1 api for eraser, a number of additional changes
had to be made.

Our eraser-tooling image had to be modified to include the
`conversion-gen` tool. This is to generate type conversions between the
older `v1alpha1` API and the newer `v1`.

To facilitate conversion between the two, an `unversioned` api was
created. This becomes a common hub between `v1` and `v1alpha1` and
facilitates conversion. It also makes it easier for our code: with the
`unversioned` api, we no longer need to update large swaths of code when
a new api version is introduced.

In order to get the `conversion-gen` tool to work, I needed to add
several magic comments throughout the api code. Most importantly, I
needed to add a `doc.go` file to each of the api version directories.
`conversion-gen` looks for this file and will not work without it. Above
the package name in those files, I added the following two magic
comments:

```
+// +k8s:conversion-gen=github.com/Azure/eraser/api/unversioned
+// -external-types=github.com/Azure/eraser/api/v1
```

This tells `conversion-gen` to use the `unversioned` api as the
conversion hub.

To skip kubebuilder processing the `unversioned` api,
`+kubebuilder:skip` had to be added before the package declaration. To
specify which api version should be stored in etcd, the
`+kubebuilder:storageversion` comment was added to the `v1` types. In
addition, `kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated"`
was added to the `v1alpha1` API in order to warn the user to upgrade.

In the `groupversion_info.go` file of each version directory, I had to
add the line

```go
localSchemeBuilder = runtime.NewSchemeBuilder(SchemeBuilder.AddToScheme)
```

To be perfectly honest, I'm not sure what this does, but it is required
for the `conversion-gen` process to generate working code.

Finally, a new file was added to `test/e2e/test-data` which deploys an
imagelist using the new API version. One of the tests was updated to use
this file instead.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on Makefile:
```diff
@@ -107,8 +107,13 @@ manifests: __controller-gen ## Generates k8s yaml for eraser deployment.
 		--load_restrictor LoadRestrictionsNone /eraser/third_party/open-policy-agent/gatekeeper/helmify | go run third_party/open-policy-agent/gatekeeper/helmify/*.go
 
-generate: __controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
+generate: __conversion-gen __controller-gen ## Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
```
can we update the comment

pmengelbert replied:
done

sozercan commented on api/unversioned/imagejob_types.go:
```diff
@@ -0,0 +1,95 @@
+	Status ImageJobStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:skip
```
i think we only need to declare this once

pmengelbert replied:
done

sozercan commented on test/e2e/test-data/eraser_v1_imagelist.yaml:
```diff
@@ -0,0 +1,9 @@
+apiVersion: eraser.sh/v1
```
let's keep this in alpha to make sure it's compatible

pmengelbert replied:
For this one, I added a separate file with v1, and modified only one of the tests to make sure that v1 works. All the other tests that use imagelists directly are using v1alpha1

sozercan commented on build/tooling/Dockerfile:
```diff
@@ -1,5 +1,6 @@
 FROM golang:1.18-bullseye
```
```suggestion
FROM golang:1.19-bullseye
```

pmengelbert replied:
done

sozercan commented on Makefile:
```diff
@@ -107,8 +107,13 @@ manifests: __controller-gen ## Generates k8s yaml for eraser deployment.
 	$(CONTROLLER_GEN) object:headerFile="hack/boilerplate.go.txt" paths="./..."
+	$(CONVERSION_GEN) \
+		--output-base=/eraser \
+		--input-dirs=./api/v1,./api/v1alpha1,./api/unversioned \
```
```suggestion
		--input-dirs=./api/... \
```

pmengelbert replied:
done

pmengelbert replied:
Not sure why, but when I do this I get the error
```
docker run --rm -v /opt/src/go/eraser:/eraser eraser-tooling conversion-gen \
        --output-base=/eraser \
        --input-dirs=./api/... \
        --go-header-file=./hack/boilerplate.go.txt \
        --output-file-base=zz_generated.conversion
W1216 16:01:48.513028       1 parse.go:247] Ignoring directory ./api: No files for pkg "./api"

```

sozercan replied:
i think you need an empty package for this to work like https://github.com/open-policy-agent/frameworks/blob/master/constraint/pkg/apis/externaldata/group.go

pmengelbert replied:
done

sozercan commented on api/v1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+limitations under the License.
+*/
+
+// Package v1 contains API Schema definitions for the batch v1 API group
```
```suggestion
// Package v1 contains API Schema definitions for the eraser.sh v1 API group
```

update in v1alpha1 too

pmengelbert replied:
done

sozercan commented on api/v1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+/*
```
let's add deprecated marker to v1alpha1 (commenting here since those files are not changed)

pmengelbert replied:
done

sozercan commented on test/e2e/tests/imagelist_exclusion_list/eraser_test.go:
```diff
@@ -91,7 +91,7 @@ func TestExclusionList(t *testing.T) {
 
 			// create imagelist to trigger deletion
-			if err := util.DeployEraserConfig(cfg.KubeconfigFile(), cfg.Namespace(), "../../test-data", "eraser_v1alpha1_imagelist.yaml"); err != nil {
+			if err := util.DeployEraserConfig(cfg.KubeconfigFile(), cfg.Namespace(), "../../test-data", "eraser_v1_imagelist.yaml"); err != nil {
```
```suggestion
			if err := util.DeployEraserConfig(cfg.KubeconfigFile(), cfg.Namespace(), "../../test-data", "eraser_v1alpha1_imagelist.yaml"); err != nil {
```

sozercan commented on api/v1/imagelist_types.go:
```diff
@@ -0,0 +1,62 @@
+/*
```
v1 and v1alpha1 should have zz_conversion files. make sure to add the conversion-gen markers and `make generate`

pmengelbert replied:
done

sozercan commented on controllers/imagecollector/imagecollector_controller.go:
```diff
@@ -233,14 +233,14 @@ func (r *Reconciler) createImageJob(ctx context.Context, req ctrl.Request, argsC
 	startTime = time.Now()
 
-	job := &eraserv1alpha1.ImageJob{
+	job := &eraserv1.ImageJob{
```
let's keep this on v1alpha1 to make sure it works too

pmengelbert replied:
I'm fine making this change, just want to understand why. If we're creating a new API version, why use the old version in the code?

sozercan replied:
I think we just want a few remaining to ensure we haven't broken existing usage. I am okay to keep this in v1 though since we already have a few scattered around.

sozercan commented on api/unversioned/imagelist_types.go:
```diff
@@ -0,0 +1,62 @@
+	Items           []ImageList `json:"items"`
+}
+
+func (*ImageListList) Hub() {}
```
i don't think we need these?

pmengelbert replied:
done

sozercan commented on api/unversioned/imagejob_types.go:
```diff
@@ -0,0 +1,95 @@
+	Digests []string `json:"digests,omitempty"`
+}
+
+func (*Image) Hub() {}
```
we don't need these

pmengelbert replied:
done

pmengelbert committed:
Re-run generation and make manifests

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove unused lines

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
For now, use --server-side to circumvent CRD size problems

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Rebase on PodTemplate changes (#555) and regenerate

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove ImageJobSpec everywhere

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Address linter concerns

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Add empty package

In order to avoid changing the Makefile every time  a new version is
added, add an empty package to the `./api` directory in order to use
the `...` wildcard.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert commented:
Rebased on main and ready for review!

pmengelbert committed:
Remove --server-side flag from Makefile

This flag will no longer be needed, because we have reduced the size of
the CRDs.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on Makefile:
```diff
@@ -205,7 +212,7 @@ deploy: manifests ## Deploy controller to the K8s cluster specified in ~/.kube/c
 	docker run --rm -v $(shell pwd)/config:/config \
 		k8s.gcr.io/kustomize/kustomize:v${KUSTOMIZE_VERSION} build \
-		/config/default | kubectl apply -f -
+		/config/default | kubectl apply --server-side -f -
```
we should be okay to drop --server-side now?

pmengelbert replied:
Whoops, I removed another but missed this one.

Done

sozercan commented on api/v1alpha1/imagejob_types.go:
```diff
@@ -62,6 +62,7 @@ type ImageJobStatus struct {
 // +kubebuilder:object:root=true
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:scope="Cluster"
+// +kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated"
```
```suggestion
// +kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated. Please migrate to v1."
```

pmengelbert committed:
Remove --server-side flag in Makefile (again)

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Instruct user to migrate to v1 in deprecation warning

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan commented on api/v1alpha1/imagelist_types.go:
```diff
@@ -38,6 +38,7 @@ type ImageListStatus struct {
 // +kubebuilder:object:root=true
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:scope="Cluster"
+// +kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated"
```
```suggestion
// +kubebuilder:deprecatedversion:warning="v1alpha1 of the eraser API has been deprecated. Please migrate to v1."
```

sozercan reviewed (approved):
Thank you! LGTM

sozercan commented:
looks like we might need a final `make manifests`

pmengelbert committed:
Re-generate manifests and codegen after rebasing

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>
